core/embed/rust/src/trezorhal/storage_field.rs

```diff
@@ -0,0 +1,379 @@
+    fn delete(&self) -> StorageResult<()>;
+}
+impl<T> FieldOpsBase for Field<T> {
+    fn is_set(&self) -> bool {
+        self.has()
+    }
+
+    fn delete(&self) -> StorageResult<()> {
+        self.delete()
+    }
+}
+
+pub trait FieldGetSet<T> {
+    fn get(&self) -> Option<T>;
+    fn set(&self, value: T) -> StorageResult<()>;
+}
+
+pub trait FieldOps<T>: FieldOpsBase + FieldGetSet<T> {}
+impl<T> FieldOps<T> for Field<T> where Field<T>: FieldGetSet<T> {}
+
+impl FieldGetSet<u32> for Field<u32> {
+    fn get(&self) -> Option<u32> {
+        let mut buf = [0u8; 4];
+        let len = storage::get(self.appkey(), &mut buf);
+        if matches!(len, Ok(4)) {
+            Some(u32::from_le_bytes(buf))
+        } else {
+            None
+        }
+    }
+
+    fn set(&self, val: u32) -> StorageResult<()> {
+        storage::set(self.appkey(), &val.to_le_bytes())
+    }
+}
+
+impl FieldGetSet<u16> for Field<u16> {
+    fn get(&self) -> Option<u16> {
+        let mut buf = [0u8; 2];
+        let len = storage::get(self.appkey(), &mut buf);
+        if matches!(len, Ok(2)) {
+            Some(u16::from_le_bytes(buf))
+        } else {
+            None
+        }
+    }
+
+    fn set(&self, val: u16) -> StorageResult<()> {
+        storage::set(self.appkey(), &val.to_le_bytes())
+    }
+}
+
+impl FieldGetSet<u8> for Field<u8> {
+    fn get(&self) -> Option<u8> {
+        let mut buf = [0u8];
+        let len = storage::get(self.appkey(), &mut buf);
+        if matches!(len, Ok(1)) {
+            Some(buf[0])
+        } else {
+            None
+        }
+    }
+
+    fn set(&self, val: u8) -> StorageResult<()> {
+        storage::set(self.appkey(), &[val])
+    }
+}
+
+impl FieldGetSet<bool> for Field<bool> {
+    fn get(&self) -> Option<bool> {
+        let mut buf = [0u8];
+        let len = storage::get(self.appkey(), &mut buf);
+        if matches!(len, Ok(1)) {
+            Some(buf[0] == TRUE_BYTE)
+        } else {
+            None
+        }
+    }
+
+    fn set(&self, val: bool) -> StorageResult<()> {
+        storage::set(self.appkey(), &[if val { TRUE_BYTE } else { FALSE_BYTE }])
+    }
+}
+
+/// Helper function defined only on boolean Fields
+pub fn set_true_or_delete(field: impl FieldOps<bool>, value: bool) -> StorageResult<()> {
+    if value {
+        field.set(true)
+    } else {
+        field.delete()
+    }
+}
+
+impl<const N: usize> FieldGetSet<String<N>> for Field<String<N>> {
+    fn get(&self) -> Option<String<N>> {
+        let mut buf = [0u8; MAX_STRING];
+        let len = storage::get(self.appkey(), &mut buf);
+        len.ok().and_then(|len| {
+            let string = String::from(str::from_utf8(&buf[..len as usize]).unwrap());
+            if self.is_len_ok(string.chars().count(), N) {
+                Some(string)
+            } else {
+                None
+            }
+        })
+    }
+
+    fn set(&self, val: String<N>) -> StorageResult<()> {
+        if self.is_len_ok(val.chars().count(), N) {
```

Comment:
Chars vs. bytes

Reply:
Changed to `val.len()`.

It is a question if this scenario can even happen, my guess is probably not, as `String<N>` needs to be already created, and if it is longer than `N`, then it will fail to create - with `String is too long to fit` error from `buffer.rs`.

Are we safe to delete the `is_len_ok` check for both the `String<N>` and `Vec<u8, N>`?

Reply:
Yes, `String<N>` is guaranteed to be at most `N` bytes long, though `is_len_ok` checks for exact length in some cases so we should leave it?